### ISSUE TITLE
 createJSON error: Error in term.topic.frequency[as.matrix(tinfo[c("Category", "Term")])] : subscript out of bounds

### ISSUE THREAD

**Issue by sdoyen** [cpsievert/LDAvis#23]
Hi,

First off, congratulation on the amazing work on LDAvis!

I'm trying to use this package to explore themes in a set of tweets (see dataset attached https://s3.amazonaws.com/rtmpre8k2d5awybtxxod-segue/tweets.csv) using the code below.

Yet I'm facing the following problem:
The createJSON function returns the following error: Error in term.topic.frequency[as.matrix(tinfo[c("Category", "Term")])] : subscript out of bounds

Any idea on what the issue could be?
Thanks in advance for your answer.

```{r}
library(LDAvis)
library(tm)
library(lda)

tweets.raw <- read.csv("/Rwd/tweets.csv")

txt <- as.list(as.character(tweets.raw$contents))
nms <- as.list(as.character(tweets.raw$authname))
nms <- gsub("[^[:graph:]]", " ",nms)
tweets <- setNames(txt, nms)
tweets <- sapply(tweets, function(x) paste(x, collapse = " "))

stop_words <- stopwords("SMART")

# tweets prepartaion
tweets <- gsub("'", "", tweets)  
tweets <- gsub("[[:punct:]]", " ", tweets)  
tweets <- gsub("[[:cntrl:]]", " ", tweets) 
tweets <- gsub("[^[:graph:]]", " ",tweets)
tweets <- tolower(tweets) 

# tokenize
doc.list <- strsplit(tweets, "[[:space:]]+")

#table of terms
term.table <- table(unlist(doc.list))
term.table <- sort(term.table, decreasing = TRUE)

#remove terms that are stop words or occur fewer than 5 times:
del <- names(term.table) %in% stop_words | term.table < 5
term.table <- term.table[!del]
vocab <- names(term.table)

#format to lda package:
get.terms <- function(x) {
  index <- match(x, vocab)
  index <- index[!is.na(index)]
  rbind(as.integer(index - 1), as.integer(rep(1, length(index))))
}
documents <- lapply(doc.list, get.terms)

#the data set stats
D <- length(documents) 
W <- length(vocab)  
doc.length <- sapply(documents, function(x) sum(x[2, ]))  
N <- sum(doc.length) 
term.frequency <- as.integer(term.table)

# MCMC and model tuning parameters:
K <- 20
#G <- 5000
G <- 100
alpha <- 0.02
eta <- 0.02

# Fit the model:
set.seed(357)
t1 <- Sys.time()
fit <- lda.collapsed.gibbs.sampler(documents = documents, K = K, vocab = vocab, 
                                   num.iterations = G, alpha = alpha, 
                                   eta = eta, initial = NULL, burnin = 0,
                                   compute.log.likelihood = TRUE)
t2 <- Sys.time()
t2 - t1  

theta <- t(apply(fit$document_sums + alpha, 2, function(x) x/sum(x)))
phi <- t(apply(t(fit$topics) + eta, 2, function(x) x/sum(x)))

tweets.list <- list(phi = phi,
                     theta = theta,
                     doc.length = doc.length,
                     vocab = vocab,
                     term.frequency = term.frequency)

#create the JSON object to feed the visualization:
json <- createJSON(phi = tweets.list$phi, 
                   theta = tweets.list$theta, 
                   doc.length = tweets.list$doc.length, 
                   vocab = tweets.list$vocab, 
                   term.frequency = tweets.list$term.frequency)

#serVis(json, out.dir = 'vis', open.browser = FALSE)
```

**Comment by cpsievert** [comment]
It looks as though you have a blank token.

```r
head(vocab, 10)
#>  [1] "http"       "coffee"     "rt"         "iced"       "latte"      "espresso"  
#>  [7] ""           "cappuccino" "black"      "starbucks" 
```

I'm able to fix your problem by simply adding an empty character string to the list of stop words. That is,

```r
stop_words <- c(stopwords("SMART"), "")
```

I'm not sure if this should be considered a bug in LDAvis. Hey @kshirley, do you think it makes sense to support empty character strings as tokens?

**Comment by kshirley** [comment]
I think we should probably throw an error if any of the terms in the vocabulary have zero characters. Even if a user wants to count spaces as tokens, for example (just trying to think of a possible explanation for the blank term), it's probably easier to denote them with a character string like "#space" in the vocabulary.

@sdoyen, out of curiosity, did the blank term in your data signify something? Or was this just a text processing issue (I know twitter data can be pretty messy)...

Carson, if you agree, then I think this will be a one-line fix in createJSON() where we just add a check on nchar(vocabulary).

**Comment by sdoyen** [comment]
Thanks for your swift and very helpful reply.

The blank token is indeed a genuine text processing issue and should't be taken into account in the analysis.

Though I would not recommend the use of "#space" which might look like a twitter hashtag, I think that createJSON() could benefit from blank space and illegal characters checking.

Thanks again.

**Comment by kshirley** [comment]
Alright, I just pushed a new version that checks that every element of the vocabulary has at least one character.

I also ran your code, @sdoyen, and noticed that we left out an important step in our "reviews" example: stripping whitespace at the beginning and end of each document. This appears to be what caused the occurrence of a blank token in some of your documents. I added two lines to the reviews example to fix this:

```r
reviews <- gsub("^[[:space:]]+", "", reviews) # remove whitespace at beginning of documents
reviews <- gsub("[[:space:]]+$", "", reviews) # remove whitespace at end of documents
```

Good point about the hashtag -- sometimes with non-twitter data I use "#" to denote a class of terms, like "#datestamp" or "#url" to capture all dates, URLs, etc. That's what I was saying one could do if there was some special class of whitespace characters that a user wanted to represent with a token. But yeah, w/ twitter data it would be confusing...

Anyway, looks like cool data (and lots of people discussing coffee in different ways!) Thx for trying LDAvis.

cheers,
k
